src/hmem_ipc_cache.c

```diff
@@ -0,0 +1,164 @@
+		 * the dead_region_list via `until_mr_uncache_entry`.
+		 * We need to flush the cache to purge the entries
+		 * and close the handles.
+		 */
+		ofi_mr_cache_flush(cache, false);
+		ret = ofi_hmem_open_handle(entry->info.iface, (void **)&entry->info.ipc_handle,
+						entry->info.device, &entry->info.ipc_mapped_addr);
+	}
+	if (ret) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to open hmem handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+	}
+	return ret;
+}
+
+static void ipc_cache_delete_region(struct ofi_mr_cache *cache,
+				     struct ofi_mr_entry *entry)
+{
+	ofi_hmem_close_handle(entry->info.iface,
+			      entry->info.ipc_mapped_addr);
+}
+
+/**
+ * @brief Open an ipc cache
+ * 
+ * @param cache[in] the ipc cache
+ * @param domain[in] the domain that the cache is attached to.
+ * @param iface[in] the hmem iface of the ipc
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_open(struct ofi_mr_cache **cache,
+			struct util_domain *domain,
+			enum fi_hmem_iface iface)
+{
+	struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {0};
+	int ret;
+
+	memory_monitors[iface] = ipc_monitor;
+	*cache = calloc(1, sizeof(*(*cache)));
+	if (!*cache) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	(*cache)->add_region = ipc_cache_add_region;
+	(*cache)->delete_region = ipc_cache_delete_region;
+	ret = ofi_mr_cache_init(domain, memory_monitors,
+				*cache);
+	if (ret)
+		goto cleanup;
+
+	FI_INFO(&core_prov, FI_LOG_CORE,
+		"ipc cache enabled, max_cnt: %zu max_size: %zu\n",
+		cache_params.max_cnt, cache_params.max_size);
+	return FI_SUCCESS;
+
+cleanup:
+	free(*cache);
+	*cache = NULL;
+out:
+	return ret;
+}
+
+/**
+ * @brief Destroy the ipc cache
+ * 
+ * @param cache the ipc cache
+ */
+void ofi_ipc_cache_destroy(struct ofi_mr_cache *cache)
+{
+	ofi_mr_cache_cleanup(cache);
+	free(cache);
+	cache = NULL;
+}
+
+/**
+ * @brief Given ipc_info (with ipc_handle and the iov of the device allocation),
+ * assign the mapped_addr the mapped address of the ipc_handle.
+ * Each (ipc_handle, mapped_addr) pair is stored in ofi_mr_entry.info.ipc_info as
+ * part of each mr entry.
+ * In a cache hit, the mapped_addr is retrieved from the matched mr entry. Otherwise,
+ * the mapped_addr is obtained by opening the ipc handle.
+ * 
+ * @param[in] cache the ipc cache
+ * @param[in] ipc_info the information of the ipc to be mapped.
+ * @param[out] mr_entry the matched mr_entry of the ipc_info and mapped_addr.
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_search(struct ofi_mr_cache *cache, struct ipc_info *ipc_info,
+			  struct ofi_mr_entry **mr_entry)
+{
+	struct ofi_mr_info info;
+	struct ofi_mr_entry *entry;
+	int ret;
+
+	info.iov.iov_base = (void *) (uintptr_t) ipc_info->base_address;
+	info.iov.iov_len = ipc_info->base_length;
+	info.iface = ipc_info->iface;
+	memcpy(&info.ipc_handle, &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
```

Comment:
Add an ipc_handle_size field to the cache and optimize the memcpy and memcmp routines.  This will ensure that we don't access data beyond the array that's valid and provide iface specific optimizations.

Reply:
How about we adding `ipc_handle_size` as part of `ipc_info`, and `ofi_mr_info`? Then we will do
```
memcpy(&info.ipc_handle, &ipc_info->ipc_handle, ipc_info->ipc_handle_size);
```
here,  and 

```
memcmp((void **)&info->ipc_handle,
		(void **)&entry->info.ipc_handle,
		entry->info.ipc_handle_size) == 0;
```

in `valid()`.